src/libraries/System.Private.CoreLib/src/System/Threading/WaitSubsystem.WaitableObject.Unix.cs

```diff
@@ -506,27 +504,22 @@ public int Wait_Locked(ThreadWaitInfo waitInfo, int timeoutMilliseconds, bool in
                     }
 
                     waitableObjects = null; // no need to clear this anymore, RegisterWait / Wait will take over from here
+
                     waitInfo.RegisterWait(count, prioritize, waitForAll);
-                    needToWait = true;
+                    return waitInfo.Wait(timeoutMilliseconds, interruptible, isSleep: false, ref lockHolder);
                 }
                 finally
                 {
+                    lockHolder.Dispose();
```

Comment:
Should this be after clearing waitableObjects?

Reply:
`waitableObjects` is thread local pooled array. It should be safe to clean it outside the lock.